beacon-chain/beacon-state-dht.md

```diff
@@ -0,0 +1,81 @@
+1. `content_key`
+1. `content_id` 
+1. `payload`
+
+The content of the message is a Merkle proof contains multiple leave nodes for a [BeaconState](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/beacon-chain.md#beaconstate).
+
+```python
+class BeaconState(Container):
+    """
+    See: https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/beacon-chain.md#beaconstate
+    """
+    pass
+
+# Used to limit the size of the message. This could be modified to be higher if needed.
+LEAVE_ITEM_LIMIT = 128
+
+# The maximum number of proof nodes needed to complete a merkle proof
+# Note that max_chuck_count() is recursively apply chunk_count() on all possible paths of the BeaconState.
+HELPER_INDICES_LIMIT = LEAVE_ITEM_LIMIT * log(max_chuck_count(BeaconState))  
+
+class BeaconStateProof(Container):
+    """
+    See: https://github.com/ethereum/consensus-specs/blob/dev/ssz/merkle-proofs.md#merkle-multiproofs
+    """
+    root: Bytes323
+    leave_indices: List[unit64, LEAVE_ITEM_LIMIT]  # GeneralizedIndex is represented by unit64
+    leaves: List[Bytes32, LEAVE_ITEM_LIMIT]
+    proof_nodes: List[Bytes32, HELPER_INDICES_LIMIT]
+```
+
+Finally, we define the necessary encodings. A light client only knows the root of the beacon state. The client wants to know the details of some leave nodes. The client has to be able to construct the `content_key` only knowing the root and which leave nodes it wants see. The `content_key` is the ssz serialization of the paths. The paths represent the part of the beacon state that one wants to know about. The paths are represented by generalized indices. Note that `hash_tree_root` and `serialize` are the same as those defined in [sync-gossip](sync-gossip.md). 
+
+```python
+class BeaconStateProofKey(Container):
+    root: Bytes323
+    leave_indices: List[unit64, LEAVE_ITEM_LIMIT]  # GeneralizedIndex is represented by unit64
+
+
+def get_generalized_index(typ: SSZType, path: Sequence[Union[int, SSZVariableName]]) -> GeneralizedIndex:
+    """
+    Converts a path (eg. `[7, "foo", 3]` for `x[7].foo[3]`, `[12, "bar", "__len__"]` for
+    `len(x[12].bar)`) into the generalized index representing its position in the Merkle tree.
+
+    See: https://github.com/ethereum/consensus-specs/blob/dev/ssz/merkle-proofs.md#ssz-object-to-index
+    """
+    pass
+
+
+def get_content_key(root: Bytes323, paths: Sequence[Sequence[[Union[int, SSZVariableName]]]) -> bytes:
+    indices = [get_generalized_index(BeaconSate, path) for path in paths]
+    proof_key = BeaconStateProofKey()
+    proof_key.root = root
+    proof_key.leave_indices = indices
+    return serialize(BeaconStateProofKey, proof_key)
+
+
+content_key := get_content_key(root, paths)
+content_id := hash_tree_root(BeaconStateProof, beacon_state_proof)
+payload := serialize(BeaconStateProof, beacon_state_proof)
+```
+
+
+## TODOs
+- Determine if the message type `BeaconStateProof` needs to be gossiped as well.
```

Comment:
Each participant of the network will be interested in different parts of the state. Those tracking L1 data would want [`latest_execution_payload_header`](https://github.com/ethereum/consensus-specs/blob/dev/specs/merge/beacon-chain.md#beaconstate) while sharded L2s would want [`shard_buffer`](https://github.com/ethereum/consensus-specs/blob/dev/specs/sharding/beacon-chain.md#beaconstate)

Reply:
Yeah...we can start this message type as a DHT network, and if the request frequency is high and the data recency is important for some use cases for specific part of the state, we could propose a "gossip" network for those data type.

Question, what kind of light client will need constant update of `latest_execution_payload_header`? What purposes do these light clients serve?

Reply:
> Question, what kind of light client will need constant update of latest_execution_payload_header? What purposes do these light clients serve?

Any client wanting to track what we currently call Ethereum mainnet head, after the merge would need to track the beacon chain head and get a `latest_execution_payload_header` proof for every block to track the execution header (current Ethereum mainnet)

Reply:
Yup. That is correct. 

I was wondering what is the end user story about this "light client"? Could this be done on demand instead? 
This might clarifies why I am asking "what kind of clients". https://ethresear.ch/t/beacon-chain-light-client-classification/11061